torchao/_models/llama/generate.py

```diff
@@ -297,6 +296,29 @@ def main(
             dtype = _NBITS_TO_DTYPE[nbits]
             group_size = int(_quant_args[2])
             quantize_(model, uintx_weight_only(dtype, group_size, use_hqq=use_hqq))
+        elif "int8_dynamic_activation_intx_weight" in quantization:
+            from torchao.experimental.quant_api import int8_dynamic_activation_intx_weight
+            assert precision == torch.float32, "int8_dynamic_activation_intx_weight requires fp32 precision"
+
+            # Build kernels in temp location, and load them in torch
+            # This requires an ARM CPU
+            from torchao.experimental.temp_build import temp_build_and_load_torchao_ops
+            temp_build_and_load_torchao_ops(cmake_lists_path=os.path.dirname(os.path.realpath(__file__)) + "/../../experimental")
```

Comment:
I dont follow why we are doing this way? Why can we not just try to load the ops and if not found raise exception with build steps needed? Ideally we can detect the platform and build this deps as part of some pre-req for running benchmarks under _model directory. When we are able to ship these kernels as part of pip package, we may not need this

Reply:
> When we are able to ship these kernels as part of pip package, we may not need this

what is the blocker for this?

Reply:
we need to move out of the experimental. I think we need to follow up on this. I believe we have sufficient evidence now? @supriyar ?

Reply:
> I dont follow why we are doing this way? Why can we not just try to load the ops and if not found raise exception with build steps needed? Ideally we can detect the platform and build this deps as part of some pre-req for running benchmarks under _model directory. When we are able to ship these kernels as part of pip package, we may not need this

I guess by "just try to load the ops" you mean something like this: https://github.com/pytorch/executorch/blob/main/extension/llm/custom_ops/sdpa_with_kv_cache.py#L21-L34 (but with the except block replaced by install instructions)

We could do that, but in the current setup, won't the try block always fail when running this benchmarking script (unless we build/load the ops in setup.py)?  I'm also not sure what the instructions would say to make the script runnable without telling the user to modify the script by adding a torch.load_library line.  I guess we could ask them to define an environment variable with the library location?

Reply:
> but with the except block replaced by install instructions

> I'm also not sure what the instructions would say to make the script runnable without telling the user to modify the script by adding a torch.load_library line.

@metascroy Yes, but I dont follow the second part. Why user needs to modify the script. Can we not just `import torchao.experimental.lowbit_ops` which internally does try/except? But I do kind of get what you are doing here because any build scripts will ahve to figure out where to put the build artifact (.so) and then we need to load from there.

Ideally it should be installed as part of the setup instructions or pip package. So we can also follow something like https://github.com/pytorch/ao/blob/main/setup.py#L53 and add extra option to build with experimental lowbit quant features. So if user invokes the benchmarking script without building experimental kernels you can suggest please do `python setup.py --use_experimental_low_bit_kernels` or `pip install . --use_experimental_low_bit_kernels`. This feels a bit cleaner to me but curious to know your thoughts and also from @msaroufim

Reply:
> Can we not just import torchao.experimental.lowbit_ops which internally does try/except

This is not going to know where to load the ops with torch.load_library unless it is responsible for both building and loading.  There are a few options we could do:

1) If a user tries to load the ops, ask the user to build the ops and define the location in some environment variable like TORCHAO_OP_LIB_PATH.  This is external to the benchmarking script.  

2) If user tries to load the ops, ask the user to run the build script we provide in experimental (experimental/build_torchao_ops.sh).  We can look for ops where build script places them if TORCHAO_OP_LIB_PATH is not defined.  This is close to what is currently done in torchchat.

3) Build these ops as part of torchao's setup.py.  I'm not sure how well this works with cmake, though.

Reply:
3 is my preference as I suggested earlier.

> I'm not sure how well this works with cmake, though.

Any specific concern or just general feeling?

Reply:
The concern is whether we can reuse the cmake setup stuff we already have (e.g., this function that sets up the parallel compute: https://github.com/pytorch/ao/blob/main/torchao/experimental/Utils.cmake#L7)?  If we bring in KleidiAI/CPUInfo via CMake, that will be more stuff to worry about.

I haven't used the torch CppExtension in setup.py, but it looks fairly simplistic compared to cmake.  Perhaps we could do something like what this blog does, if it does not already exist in PyTorch: https://martinopilia.com/posts/2018/09/15/building-python-extension.html

Reply:
Maybe I am missing something but Executorch's pybinding extension with xnnpack builds xnnpack which build with cpuinfo and pthreadpool and everything. Granted that is a whole lot to build but it does work in ET. So likely there is a nuance that you are worried about that I am not understanding

Reply:
I guess this goes back to one of my earlier comments to @jerryzh168: the current setup.py in torchao does not really support cmake, and it would require a good amount of refactoring to support it.  Currently setup.py in torchao is built on utilities in torch.utils.cpp_extension, which look somewhat simplistic and as far as I can tell, do not support cmake.

ET's setup.py defines a custom extension to support cmake and doing something similar to torchao looks like a sizable refactor to their setup?

Reply:
Ok I will have to rely on your answer for this since I havent looked at all the details of cpp_extension. I do remember it being simple but I dont know if there are ways to add deps as part of cpp_extension. I doubt but see if possible.

If not, I think it is worth proposing this for the sake of making ARM kernels available on mac builds. @drisspg any thoughts? discussion here is largely around more complex setup.py in order to allow us to build cpp package extensions that package cpu kernels and make it available as part of python package